src/other/index.jl

```diff
@@ -114,8 +114,8 @@ Base.haskey(x::Index, key::Integer) = 1 <= key <= length(x.names)
 Base.haskey(x::Index, key::Bool) =
     throw(ArgumentError("invalid key: $key of type Bool"))
 
-# TODO: If this should stay 'unsafe', perhaps make unexported
-function Base.push!(x::Index, nm::Symbol)
+function safe_push!(x::Index, nm::Symbol)
```

Comment:
Why not keep this a method of `Base.push!`? The fact that it checks whether the name already exists seems consistent with `Base.push!`. If anything, the old method should have been called `unsafe_push!`.

(Also I assume you're confident it doesn't affect performance too much in practice.)

Reply:
This is the overhead:
```
julia> df = DataFrame(ones(0, 10000));

julia> @benchmark hasproperty($df, :x)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.030 ns (0.00% GC)
  median time:      6.913 ns (0.00% GC)
  mean time:        7.308 ns (0.00% GC)
  maximum time:     25.230 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999

julia> @benchmark hasproperty($df, :x1000)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.279 ns (0.00% GC)
  median time:      7.212 ns (0.00% GC)
  mean time:        7.777 ns (0.00% GC)
  maximum time:     32.975 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
```
and I would say it is negligible, given we do `safe_push!` in a very specific context - i.e. creating a new column in a data frame (which in particular most likely will be copied anyway which is orders of magnitude slower).

Reply:
I would not keep it as `Base.push!` because we do not export `Index` so I would prefer not to pollute `push!` with this method. This `safe_push!` has a very specific and narrow use case. The original code AFAICT was designed when it was assumed that `Index` would be exported, but it is not (and I think it does not have to be).

Reply:
Well we don't pollute anything since it's not exported. We also define `haskey`, `merge!`, `delete!`, `empty!`, `insert!`, `getindex`, `isequal`...

Reply:
OK - I have reverted `push!`.